### PR TITLE
feat(dashboard): embedded assets into binary for release builds

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

See https://github.com/risingwavelabs/risingwave/issues/15280#issuecomment-1993795653 for the background.

Introduce a `dashboard` crate for building and embedding assets. The embedded mode and proxied mode will be chosen based on configuration.

https://github.com/risingwavelabs/risingwave/blob/1dab4105c3bb66f7a53d9b2a03539396ba3df819/src/meta/dashboard/src/lib.rs#L22-L36

Note that there's no actual change on the binary release or docker release. This PR is mainly focusing on introducing the functionality. Will deprecate the old external ui approach in next PRs.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15656?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15656** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15656?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#15655** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15655?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BugenZhao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/meta/dashboard/src/embed.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+
+use axum::http::{header, StatusCode, Uri};
+use axum::response::{Html, IntoResponse as _, Response};
+use axum::Router;
+use rust_embed::RustEmbed;
+
+#[derive(RustEmbed)]
+#[folder = "$OUT_DIR/assets"]
+struct Assets;
```
Comment (xxchan) [inline src/meta/dashboard/src/embed.rs:22]: What's the size of the embedded asset?

**Review by xxchan — APPROVED** [review]
Generally LGTM

File: src/meta/dashboard/build.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    let dest = dest_dir();
+    let src = Path::new(DASHBOARD_DIR).join("out");
+    dircpy::copy_dir(src, dest)?;
+
+    Ok(())
+}
+
+fn main() -> anyhow::Result<()> {
+    rerun_if_changed!("build.rs");
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/meta/dashboard/build.rs:64]: 😮 TIL

> Note that if the build script itself (or one of its dependencies) changes, then it's rebuilt and rerun unconditionally, so rerun_if_changed!("build.rs") is almost always redundant (unless you want to ignore changes in all other files except for build.rs).

**Review by xxchan — APPROVED** [review]
[gogogo](https://app.graphite.dev/settings/meme-library?org=risingwavelabs)

**Pull request merged**